src/source.rs

```diff
@@ -0,0 +1,134 @@
+// Copyright 2013-2015, The Rust-GNOME Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+//! Manages available sources of events for the main loop
+
+use std::cell::RefCell;
+use std::ops::DerefMut;
+use std::mem::transmute;
+use ffi::{gboolean, gpointer, g_idle_add_full, g_timeout_add_full, g_timeout_add_seconds_full};
+use translate::ToGlib;
+
+/// Return type of idle and timeout functions.
+///
+/// In the callback, return `Continue(true)` to continue scheduling the callback
+/// in the main loop or `Continue(false)` to remove it from the main loop.
+pub struct Continue(pub bool);
+
+impl ToGlib for Continue {
+    type GlibType = gboolean;
+
+    #[inline]
+    fn to_glib(&self) -> gboolean {
+        self.0.to_glib()
+    }
+}
+
+
+// Box::into_raw stability workaround
+unsafe fn into_raw<T>(b: Box<T>) -> *mut T { transmute(b) }
+
+extern "C" fn trampoline(func: &RefCell<Box<FnMut() -> Continue + 'static>>) -> gboolean {
+    func.borrow_mut().deref_mut()().to_glib()
+}
+
+extern "C" fn destroy_closure(ptr: gpointer) {
+    unsafe {
+        // Box::from_raw API stability workaround
+        let ptr = ptr as *mut RefCell<Box<FnMut() -> Continue + 'static>>;
+        let _: Box<RefCell<Box<FnMut() -> Continue + 'static>>> = transmute(ptr);
+    }
+}
+
+const PRIORITY_DEFAULT: i32 = 0;
+const PRIORITY_DEFAULT_IDLE: i32 = 200;
+
+
+/// Adds a function to be called whenever there are no higher priority events pending to the default main loop.
+///
+/// The function is given the default idle priority, `PRIORITY_DEFAULT_IDLE`.
+/// If the function returns `Continue(false)` it is automatically removed from
+/// the list of event sources and will not be called again.
+///
+/// # Examples
+///
+/// ```
```

Comment:
Could you also please mark the examples with `ignore` so `cargo test` won't complain that they don't compile (or alternatively make them compile)?
http://doc.rust-lang.org/book/documentation.html#running-documentation-tests